packages/plugin-data/src/api/data/tree/exportApi.ts

```diff
@@ -0,0 +1,607 @@
+    } catch (err) {
+      if (err.name === 'MALFORMED_QUERY') {
+        const errMsg = messages.getMessage('soqlMalformed');
+        const errMsgAction = messages.getMessage('soqlMalformedAction');
+        throw new SfdxError(errMsg, 'MalformedQuery', [errMsgAction]);
+      } else {
+        throw err;
+      }
+    }
+
+    const sobjectTree = await this.processQueryResults(queryResults);
+
+    if (!sobjectTree.records?.length) {
+      return sobjectTree;
+    }
+    if (plan) {
+      return this.generateDataPlan(sobjectTree);
+    } else {
+      const fileName = `${Object.keys(this.objectTypeRegistry).join('-')}.json`;
+      return this.writeDataFileSync(fileName, sobjectTree);
+    }
+  }
+
+  //   P R I V A T E   M E T H O D S
+
+  /**
+   * Ensures a valid query is defined in the export configuration,
+   * which can be either a soql query or a path to a file containing
+   * a soql query.
+   *
+   * @param config - The export configuration.
+   */
+  private validate(config: ExportConfig): ExportConfig {
+    if (!config.query) {
+      throw SfdxError.create('@salesforce/plugin-data', 'exportApi', 'queryNotProvided');
+    }
+
+    const filepath = path.resolve(process.cwd(), config.query);
+    if (fs.existsSync(filepath)) {
+      config.query = fs.readFileSync(filepath, 'utf8');
+
+      if (!config.query) {
+        throw SfdxError.create('@salesforce/plugin-data', 'exportApi', 'queryNotProvided');
+      }
+    }
+
+    config.query = config.query.toLowerCase().trim();
+    if (!config.query.startsWith('select')) {
+      throw SfdxError.create('@salesforce/plugin-data', 'exportApi', 'soqlInvalid', [config.query]);
+    }
+
+    return config;
+  }
+
+  private setupOutputDirectory(outputDir: string): void {
+    try {
+      fs.mkdirpSync(outputDir);
+    } catch (err) {
+      // It is ok if the directory already exist
+      if (err.name !== 'EEXIST') {
+        throw err;
+      }
+    }
+  }
+
+  // Process query results generating SObject Tree format
+  private async processQueryResults(recordList: any): Promise<any> {
+    await this.recordObjectTypes(recordList);
+
+    const { plan, query } = this.config;
+
+    const processedRecordList = await this.processRecordList(recordList);
+    // log record count; warn if > 200 and !options.plan
+    const recordCount = getNumber(processedRecordList, 'records.length', 0);
+    this.logger.debug(messages.getMessage('dataExportRecordCount', [recordCount, query]));
+    if (recordCount > 200 && !plan) {
+      this.ux.warn(messages.getMessage('dataExportRecordCountWarning', [recordCount, query]));
+    }
+    return this.finalApplyRefs(processedRecordList.records);
+  }
+
+  // Register object types and type hierarchy for plan generation
+  private async recordObjectTypes(recordList: any): Promise<any> {
+    const records = recordList.records;
+
+    if (!records.length) {
+      // TODO: should be on the command
```

Comment:
why don't we do this now?

Reply:
You're right, and I tried to but it was causing lots of issues so I punted.  There are more work items to handle things like this so unless we really need to I'd prefer to defer.